core/src/main/java/org/infinispan/container/BoundedSegmentedDataContainer.java

```diff
@@ -0,0 +1,255 @@
+      return entries.values().spliterator();
+   }
+
+   @Override
+   public Spliterator<InternalCacheEntry<K, V>> spliteratorIncludingExpired(IntSet segments) {
+      // Copy the ints into an array to parallelize them
+      int[] segmentArray = segments.toIntArray();
+      AtomicBoolean usedOthers = new AtomicBoolean(false);
+
+      return new ComposeArraySpliterator<>(i -> {
+         ConcurrentMap<K, InternalCacheEntry<K, V>> map = maps.get(segmentArray[i]);
+         if (map == null) {
+            if (!usedOthers.getAndSet(true)) {
+               return entries.values().stream()
+                     .filter(e -> segments.contains(getSegmentForKey(e.getKey())))
+                     .collect(Collectors.toSet());
+            }
+            return Collections.emptyList();
+         }
+         return map.values();
+      }, segmentArray.length, Spliterator.CONCURRENT | Spliterator.NONNULL | Spliterator.DISTINCT);
+   }
+
+   @Override
+   public int sizeIncludingExpired() {
+      return entries.size();
+   }
+
+   /**
+    * Clears entries out of caffeine map by invoking remove on iterator. This can either keep all keys that match the
+    * provided segments when keepSegments is <code>true</code> or it will remove only the provided segments when
+    * keepSegments is <code>false</code>.
+    * @param segments the segments to either remove or keep
+    * @param keepSegments whether segments are kept or removed
+    */
+   private void clear(IntSet segments, boolean keepSegments) {
+      for (Iterator<K> keyIterator = entries.keySet().iterator(); keyIterator.hasNext(); ) {
+         K key = keyIterator.next();
+         int keySegment = getSegmentForKey(key);
+         if (keepSegments != segments.contains(keySegment)) {
+            keyIterator.remove();
+         }
+      }
+   }
+
+   @Override
+   public void removeSegments(IntSet segments) {
+      // Call super remove segments so the maps are remove more efficiently
+      super.removeSegments(segments);
+      // Finally remove the entries from bounded cache
+      clear(segments, false);
+   }
+
+   private Policy.Eviction<K, InternalCacheEntry<K, V>> eviction() {
+      if (evictionCache != null) {
+         Optional<Policy.Eviction<K, InternalCacheEntry<K, V>>> eviction = evictionCache.policy().eviction();
+         if (eviction.isPresent()) {
+            return eviction.get();
+         }
+      }
+      throw new UnsupportedOperationException();
+   }
+
+   @Override
+   public long capacity() {
+      Policy.Eviction<K, InternalCacheEntry<K, V>> evict = eviction();
+      return evict.getMaximum();
+   }
+
+   @Override
+   public void resize(long newSize) {
+      Policy.Eviction<K, InternalCacheEntry<K, V>> evict = eviction();
+      evict.setMaximum(newSize);
+   }
+
+   @Override
+   public long evictionSize() {
+      Policy.Eviction<K, InternalCacheEntry<K, V>> evict = eviction();
+      return evict.weightedSize().orElse(entries.size());
```

Comment:
Is there are a need for the `evict` variable on the above three methods?

Reply:
There isn't specifically. This is how it was before. And I believe I did it that way before, because the `eviction` method can very easily throw a `UnsupportedOperationException` and I wanted to keep the line numbers separate.

We can merge it together though to save the 3 lines if we wanted.